include/ensmallen_bits/callbacks/report.hpp

```diff
@@ -0,0 +1,558 @@
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param gradientIn Matrix that holds the gradient.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void Gradient(OptimizerType& /* optimizer */,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                const MatType& gradientIn)
+  {
+    hasGradient = true;
+    gradientNorm = arma::norm(gradientIn);
+    gradientCalls++;
+  }
+
+  /**
+   * Callback function called at any call to GradientConstraint().
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param constraint The index of the constraint;
+   * @param gradient Matrix that holds the gradient;
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void GradientConstraint(OptimizerType& optimizer,
+                          FunctionType& function,
+                          const MatType& coordinates,
+                          const size_t /* constraint */,
+                          const MatType& gradient)
+  {
+    Gradient(optimizer, function, coordinates, gradient);
+  }
+
+ private:
+  /**
+   * Helper function to print the number of function to the specified output
+   * stream.
+   *
+   * @param function The instantiated function that implements NumFunctions().
+   * @param stream The output stream.
+   */
+  template<typename FunctionType>
+  typename std::enable_if<
+      traits::HasNumFunctionsSignature<FunctionType>::value, void>::type
+  PrintNumFunctions(const FunctionType& function, std::stringstream& stream)
+  {
+    PrettyPrintElement(stream, "Num functions:", 30);
+    stream << function.NumFunctions() << std::endl;
+  }
+
+  template<typename FunctionType>
+  typename std::enable_if<
+      !traits::HasNumFunctionsSignature<FunctionType>::value, void>::type
+  PrintNumFunctions(const FunctionType& /* function */,
+                    std::stringstream& /* stream */) { }
+
+  /**
+   * Helper function to output the max-iterations to the specified output
+   * stream.
+   *
+   * @param optimizer The instantiated optimizer that implements
+   *     MaxIterations().
+   * @param stream The output stream.
+   */
+  template<typename OptimizerType>
+  typename std::enable_if<
+      traits::HasMaxIterationsSignature<OptimizerType>::value, void>::type
+  PrintMaxIterations(const OptimizerType& optimizer, std::stringstream& stream)
+  {
+    PrettyPrintElement(stream, "Max iterations:", 30);
+    stream << optimizer.MaxIterations() << std::endl;
+
+    PrettyPrintElement(stream, "Reached maximum iterations:", 30);
+    stream << std::string(optimizer.MaxIterations() == objectives.size() ?
+        "true" : "false") << std::endl;
```

Comment:
Up to you, but if the function is separable, do you think we should also print "Max. epochs" too or something like that?

Reply:
I called it `Number of epochs` - https://github.com/mlpack/ensmallen/pull/213/files#diff-5aa5b7e1523591f922934321e984ab31e0171e931ceee2ca499d993daf1e0a3cR173

But I can rename it.

Reply:
Ah, looks good to me!  Sorry if I missed that in the original commit. :+1: